File: 2024_algorithms/Data-Structures-and-Algorithms-Specialization-2023/1_Algorithmic_Toolbox/week1_programming_challenges/a02_maximum_pairwise_product.cpp
#include <cstdlib>
#include <iostream>
#include <vector>

// using namespace std;

long long maximumPairwiseProduct(std::vector<int>& nums) {
	long long ans = 0;

	if (nums.size() <= 1) {
		return 0;
	}

	for (int i=0; i<nums.size(); i++) {
		for (int j=i+1; j<nums.size(); j++) {
			ans = std::max(
				ans,
				((long long) (nums[i])) * ((long long) (nums[j]))
			);
		}
	}

	return ans;
}

int main() {
	srand (time(NULL));
	std::cout << "==> Random number: " << rand()%10 + 1 << std::endl;

	int size, val;

	std::cout << "==> Enter size of input: ";
	std::cin >> size;

	if (size < 2) {
		std::cerr << "!!! invalid size: " << size << std::endl;
		std::exit(1);
	}

	std::vector<int> nums(size);
	for (int i=0; i<size; i++) {
		std::cin >> nums[i];
	}

	std::cout << "==> Ans: " << maximumPairwiseProduct(nums) << std::endl;

	return 0;
}

File: 2024_algorithms/Data-Structures-and-Algorithms-Specialization-2023/1_Algorithmic_Toolbox/week1_programming_challenges/a02_maximum_pairwise_product.rs
use std::cmp::max;
use std::io::{stdin, stdout, Write};

fn main() {
    let mut input = String::new();
    let mut val;

    print!("==> Enter size of numbers: ");

    let _ = stdout().flush();
    input.clear();
    stdin().read_line(&mut input).expect("Failed to read line");
    let num: usize = input.trim().parse().expect("Input not an integer");

    if num < 2 {
        panic!("invalid size: {}", num);
    }

    let mut numbers: Vec<usize> = Vec::with_capacity(num);

    for _ in 0..num {
        input.clear();
        stdin().read_line(&mut input).expect("Failed to read line");
        val = input.trim().parse().expect("Input not an integer");
        numbers.push(val);
    }

    println!("==> Ans: {}", maximum_pairwise_product(&numbers));
}

fn maximum_pairwise_product(numbers: &Vec<usize>) -> u128 {
    let mut ans = 0;
    dbg!(&numbers);

    if numbers.len() < 2 {
        return 0;
    }

    numbers.iter().enumerate().for_each(|(i, v1)| {
        numbers[i + 1..].iter().for_each(|v2| {
            ans = max(ans, (*v1 as u128) * (*v2 as u128));
        });
    });

    ans
}
